### ISSUE TITLE
Make limiter as tolerant to overload as Calf Multiband Limiter

### ISSUE THREAD

**Issue by v-fox** [sadko4u/lsp-plugins#122]
I've been meaning to open this issue since discussing it in https://github.com/sadko4u/lsp-plugins/issues/71#issuecomment-585525088 but procrastinated. [CLM](https://github.com/calf-studio-gear/calf) allows arbitrary boost that finely preserves audio quality even at ridiculous levels (like +24db) but lsp-plugins-limiter-stereo starts to exponentially increase screeching just as level goes beyond 0-point regardless of settings. I do not know what trickery CLM uses but it seems very effective. Unfortunately, it's now quite unmaintained and has the [latency issue](https://github.com/calf-studio-gear/calf/issues/214). This and "stereo tools" (for tuning stereo balance) are my main reasons to not go full LSP in my setup, so it would be nice to be able to ditch them.

**Comment by sadko4u** [comment]
Could you please make a video demonstration capturing the output audio and showing audio levels, too, where you compare CLM and LSP Limiter for ridiculous cases?

**Comment by v-fox** [comment]
> Could you please make a video demonstration capturing the output audio and showing audio levels, too, where you compare CLM and LSP Limiter for ridiculous cases?

Well, I'm not sure about videos because of how much video encoding thrashes CPU which can lead to audio xruns or video stuttering with such substantial realtime load but I may try to dump output audio at equally high levels or something into files.

But it should be trivial for you to reproduce by taking both limiters, streaming music or speech from a player (via native JACK output from things like gst123, clementine/strawberry or via PA as JACK client from mpv which native JACK output seems glitchy) into them while slowly upping input gain until surpassing 0db and reaching +24db level.

**Comment by sadko4u** [comment]
Yes, I've reproduced the case on extreme settings, when the signal completely overloads the input.
For me such case is unnatural since limiter initially was designed to cut-off short peaks.
But implementing additional RMS control of the input signal could solve the screeching problem.

**Comment by v-fox** [comment]
I've made an [Audacity project with 2 +12db stereo streams of a short track](https://drive.google.com/drive/folders/1PMr_jgK27g7TQLkc7kjsAbMiYjOrSBTZ) (may be switched with "Solo" button while playing). Although music with lyrics at such level does not sound fine, non-vocal tracks should not be unpleasant for ears and speech alone should be fine.

> Yes, I've reproduced the case on extreme settings, when the signal completely overloads the input.
> For me such case is unnatural since limiter initially was designed to cut-off short peaks.
> But implementing additional RMS control of the input signal could solve the screeching problem.

Understandable. I use it as kind of realtime DRC for streams of speech with background sounds which are not normalized, so levels are all over the place and I try to squash them to be discernable on low hardware output volume which means that such overload is common.

**Comment by sadko4u** [comment]
@v-fox 
I've introduced Automated Level Regulation (ALR) feature to the limiter and pushed to the **github-issue-122** branch.
Please check it.

![image](https://user-images.githubusercontent.com/12834464/87168119-b2fdf880-c2d6-11ea-9eca-e410dcd9f4aa.png)

**Comment by v-fox** [comment]
@sadko4u Have tested with [the track from previous comparison](https://drive.google.com/drive/folders/1PMr_jgK27g7TQLkc7kjsAbMiYjOrSBTZ). Unfortunately, this branch doesn't seem to have "antisurge" plugin, so I had to resort to old gate-based pipeline.

No horrible screeching but required some tinkering. Not sure what "knee" does but with it were horrible clicking peaks:
![lsp-plugins-limiter-stereo_knee-clicks](https://user-images.githubusercontent.com/362763/87192700-7222d580-c310-11ea-9074-0f00c2b8e138.png)

Then I've got subtle clicks on volume burst from quiet background:
![lsp-plugins-limiter-stereo_clicks-on-bursts-from-quiet-background](https://user-images.githubusercontent.com/362763/87192719-81a21e80-c310-11ea-9968-fe44adca45a4.png)

Then it became more or less fine even with what's supposed to be +12 (actually, 18-6) dB overload:
![lsp-plugins-limiter-stereo_good-with-overload+12dB](https://user-images.githubusercontent.com/362763/87193141-7ef3f900-c311-11ea-8241-f00961643fe5.png)

So I ended up with this for normal use with just tuning "threshold" for loud or quiet sources:
![lsp-plugins-limiter-stereo_normal-use_v4](https://user-images.githubusercontent.com/362763/87202210-7dc6ca00-c319-11ea-830c-dcff4b8d41af.png)

This allows to even lowering threshold about 6 dB more for boosting songs without obviously screwing up their composition:
![lsp-plugins-limiter-stereo_smooth-boost-of+6dB](https://user-images.githubusercontent.com/362763/87202852-db0f4b00-c31a-11ea-8e63-c63475d5986f.png)

* Not sure why ALR attack is limited to 5ms at minimum, shouldn't it be equal to lookahead ? CML uses 1ms by default but that's impossible for ALR attack.

What about other stuff from CML ?
* Is Herm Thin mode an equivalent to ASC ?
* Isn't CML supposed to have some kind of dynamic release length activated by "min release" ?
* CML is also multiband, allowing to assign different weighting and release to bands. But that maybe superfluous even for me. Although, together with quick preset menu it's nice feature for giving quick smooth boosting or damping to especially quiet or annoying frequency ranges.

Anyway, now it's hard to discern difference between CML and this limiter. Excellent !

**Comment by sadko4u** [comment]
> Not sure why ALR attack is limited to 5ms at minimum, shouldn't it be equal to lookahead

ALR actually works as a compressor with small knee and infinite ratio. The main purpose is to track the RMS value of the signal and adjust corresponding gain reduction. I tried the compressor at exteme settings (-24 dB threshold and about 0 dB of the original signal) and with correct ALR Attack and Release it worked pretty fine.

> Not sure what "knee" does but with it were horrible clicking peaks

I think I should eliminate it. It controls the additional gain applied to gain patches that cut-off short peaks.

> Is Herm Thin mode an equivalent to ASC

No, Calf operates differently. I still can not get a clean result on it, I always get smoothed sound which I don't prefer.
'Herm Thin' is just a form of gain patch applied to the peak to reduce it.

> Isn't CML supposed to have some kind of dynamic release length activated by "min release"

I don't know how does CML work internally. I believe Markus also didn't know how the ASC does work since the documentation is so poor.

> CML is also multiband, allowing to assign different weighting and release to bands. But that maybe superfluous even for me

I'm planning to release a multiband limiter but not in 1.1.24 release.

> Anyway, now it's hard to discern difference between CML and this limiter. Excellent !

The difference is noticeable, especially when looking at the K12 meter. LSP limiter allows to gain 2-3 additional dB on the K12 meter against CLM while sounding less distorted.

**Comment by sadko4u** [comment]
Pushed some additional improvements to the Limiter. Latest commit is: ce830527379df377d42cb4fe1a91513a864cac25

Now **knee** knob allows to correct the control level produced by ALR to compute the gain reduction.

**Comment by v-fox** [comment]
> > Not sure why ALR attack is limited to 5ms at minimum, shouldn't it be equal to lookahead
> 
> ALR actually works as a compressor with small knee and infinite ratio. The main purpose is to track the RMS value of the signal and adjust corresponding gain reduction before the peak cutting algorithm is applied. I tried the compressor at exteme settings (-24 dB threshold and about 0 dB of the original signal) and with correct ALR Attack and Release it worked pretty fine.

Interesting.

>> Is Herm Thin mode an equivalent to ASC
>
> No, Calf operates differently. I still can not get a clean result on it, I always get smoothed sound which I don't prefer.
> 'Herm Thin' is just a form of gain patch applied to the peak to reduce it.

"Smooth" ? On simple or "busy", with whole bunch of background sounds over high frequency and volume ranges, sources ? I would like some smooth sound for the latter. Although, sometimes it also can be going from "smooth" to "damp".

> > Isn't CML supposed to have some kind of dynamic release length activated by "min release"
> 
> I don't know how does CML work internally. I believe Markus also didn't know how the ASC does work since the documentation is so poor.

Have you seen https://github.com/calf-studio-gear/calf/issues/167 ?

It seems ALR attack doesn't really depend on lookahead, as I assumed from CML not having separate attack controls. Putting in long ALR attack and release gave me nice crisp sound on things I've tested:
![lsp-plugins-limiter-stereo_big-boost](https://user-images.githubusercontent.com/362763/87238156-69b5c200-c418-11ea-8bd4-0dda5c8e086f.png)
View, of course, is also limited, so on such sources (that are typical for Youtube and movies with high DR audio) threshold has to trail near the boundary. But CML doesn't even visualize its work, only uses the view for setting bands, and its main knob is even more limited.

**Comment by sadko4u** [comment]
> Interesting.

The gain reduction algorithm is split into two stages:
* Automated Level Regulation (ALR) which acts like a compression with infinite ratio for purpose of estimating the smoothed gain reduction level. The gain reduction level is controlled by the envelope of the signal applied to the compressor's gain reduction curve. The smoothness of the envelope is controlled by the 'Attack' and 'Release' knobs in the ALR section. Since different signals have different envelopes with short release time, the 'Knee' knob allows to adjust the threshold of the compressor. Adjusting 'Knee' may give possibility to gain some additional decibels of loudness in the final result.
* Peak cutting algorithm which searches peaks above the threshold values and applies short gain reduction patches to the signal. The form of patch is controlled by the 'Mode' selector and it's length is controlled by the 'Attack' and 'Release' knobs. Note that 'Lookahead' also affects these values: 'Attack' can not be larger than 'Lookahead' and 'Release' can not be twice larger as 'Lookahead'. Setting these values larger than allowed automatically makes them considered to be set to maximum possible values.

> "Smooth" ? On simple or "busy", with whole bunch of background sounds over high frequency and volume ranges, sources ? I would like some smooth sound for the latter. Although, sometimes it also can be going from "smooth" to "damp".

I don't prefer Calf Limiter, it never allowed me to pump-in loudness of heavy metal tracks to the desired level like it's done on well-mastered studio mixes. It's always lacking 2-3 dB of gain.

> Have you seen [calf-studio-gear/calf#167](https://github.com/calf-studio-gear/calf/issues/167) ?

No, I don't have much time to track each issue in Calf issue tracker. Now it's more clean how it works but... Why it is not reflected in the documentation to the device which even doesn't provide the decryption of ASC acronym?

> It seems ALR attack doesn't really depend on lookahead, as I assumed from CML not having separate attack controls. Putting in long ALR attack and release gave me nice crisp sound on things I've tested:

Right, and previously I've explained why.

> View, of course, is also limited, so on such sources (that are typical for Youtube and movies with high DR audio) threshold 

I don't want to make each plugin to display the full available dynamic range of the signal. The range of 48 dB is pretty enough to operate with. If you're lacking some data below -48 dB, then just adjust the level of the input signal by scrolling 'Input' knob. I've just committed the changes which allow to set +60 dB input gain.

> But CML doesn't even visualize its work, only uses the view for setting bands, and its main knob is even more limited.

Moreover, these bands don't apply proper phase corrections which yields to spectrum distortion at split frequencies:
https://github.com/calf-studio-gear/calf/issues/217

**Comment by sadko4u** [comment]
Anyway, I think Limiter is ready to go into **devel** branch

**Comment by sadko4u** [comment]
Merged changes into **devel** branch: b6c8920453d6c6155acb2db9e5273b1b3fe34f10

**Comment by sadko4u** [comment]
Test fixes: 6cf4cefb7ccca047620dc8db670c31456f553cd5

**Comment by v-fox** [comment]
> Since different signals have different envelopes with short release time, the 'Knee' knob allows to adjust the threshold of the compressor. Adjusting 'Knee' may give possibility to gain some additional decibels of loudness in the final result.

I've been staring at limiter for hours and never noticed the new knee knob until now…

> Note that 'Lookahead' also affects these values: 'Attack' can not be larger than 'Lookahead' and 'Release' can not be twice larger as 'Lookahead'. Setting these values larger than allowed automatically makes them considered to be set to maximum possible values.

I've previously read about attack and release depending on lookahead of old, original version of the limiter but wasn't sure if lookahead has influence on ALR. By the way, could it be possible to show real attack and release values when they are limited like that, despite the knob's position ? Maybe clip their number and change its colour until tuning lookahead would allow to set them closer to knobs' positions.

> I don't prefer Calf Limiter, it never allowed me to pump-in loudness of heavy metal tracks to the desired level like it's done on well-mastered studio mixes. It's always lacking 2-3 dB of gain.

Ironic that you would disfavour smoother sound when boosting metal. For me it's the hardest thing to boost without making it smudged & screeched or over-compressed & damp. But my use-case mostly consist of boosting speech with background sounds while attempting not to distort either.

A lot of musical tracks seems to be mastered while targeting anything between 0 dB and -12 dB with or without some occasional peaks to 0 dB. I've also been trying to find a way of boosting quieter tracks in that range to 0 dB while preserving the form of louder tracks by avoiding over-compression. Not sure if limiter is an answer for that at all.

> No, I don't have much time to track each issue in Calf issue tracker. Now it's more clean how it works but... Why it is not reflected in the documentation to the device which even doesn't provide the decryption of ASC acronym?

I haven't implied that you could or should have. Just that it's likely the only thing that would come up in search but wasn't enough to get exact workings. That question was asked exactly because others couldn't figure it out either. Most projects have lacking documentation and even worse integration practices. Look at [Khronos](https://github.com/KhronosGroup) that develops all graphics and compute standards, their release management and feature integration are the worst (you would have to guess which revisions would compile with each other, let alone, actually work; half-baked branches are slammed into master at a whim) even, though, whole world depends on them. And Google projects… oof.

> I don't want to make each plugin to display the full available dynamic range of the signal. The range of 48 dB is pretty enough to operate with. If you're lacking some data below -48 dB, then just adjust the level of the input signal by scrolling 'Input' knob. I've just committed the changes which allow to set +60 dB input gain.

Why not use common zoom slider of universal view widget that would scale graph inside static frame from 12 dB to 144 dB with 83 dB by default and save customized value in configs ?

> Anyway, I think Limiter is ready to go into **devel** branch

It's certainly is very useful and superior to the original version.

**Comment by sadko4u** [comment]
> For me it's the hardest thing to boost without making it smudged & screeched or over-compressed & damp.

Maybe we're understanding different things under 'smoothed'. I mean that 'smoothed' signal is less punchy: it sounds stable but looses it's punch and, therefore, sounds not so 'alive'. It is especially noticeable how Kick and Snare drums behave for LSP vs Calf. I suspect that's because the Calf's release time is too long. But if I lower Calf's release it starts scratching the sound and gets 'not ok'.
Additionally, it can not go over +6 marker on Ardour's K12 meter without destroying the sound:
![image](https://user-images.githubusercontent.com/12834464/87250658-0d31c180-c46f-11ea-8fb1-d6c13565d2e5.png)

With LSP one I can get more 'alive' sound than with Calf one and additionally boost of the level on K12 meter:
![image](https://user-images.githubusercontent.com/12834464/87250639-e8d5e500-c46e-11ea-93eb-58c807e36f7a.png)

> Not sure if limiter is an answer for that at all.

Of course not. Many peaks should be removed before they reach the limiter that is the last thing in the chain.

> Why not use common zoom slider of universal view widget that would scale graph inside static frame from 12 dB to 144 dB

Because it requires additional UI tunings which are not always trivial. Each line on the graph is individual widget which requires control. And, again, I don't find it useful. If you use limiter to control more than 48 dBFS gain range, then, probably, there's something wrong with your mix and it should be updated before you apply the limiter.

**Comment by sadko4u** [comment]
Removed **github-issue-122** branch. All changes are available in **devel** now.

**Comment by sadko4u** [comment]
@wwmm @Digitalone1

Please note that this change may break current wrappings of PuleEffects around Limiter.
You can detect the new limiter plugin by it's version number which will be changed from 1.0.2 to 1.0.3 (can be obtained from LV2 metadata) or by new appeared port bindings - **alr**, **alr_at** and **alr_rt**.

**Comment by Digitalone1** [comment]
No worries, limiter plugin used in PulseEffects is the one by Calf.

**Comment by v-fox** [comment]
> > For me it's the hardest thing to boost without making it smudged & screeched or over-compressed & damp.
> 
> Maybe we're understanding different things under 'smoothed'. I mean that 'smoothed' signal is less punchy: it sounds stable but looses it's punch and, therefore, sounds not so 'alive'. It is especially noticeable how Kick and Snare drums behave for LSP vs Calf. I suspect that's because the Calf's release time is too long. But if I lower Calf's release it starts scratching the sound and gets 'not ok'.

Well, yes and no: less punch is not good but all my attempts to minimize release and allow limiter's curve to be less smoothed always lead to some kind of scratchiness. So, less punch is not good but it's, at least, "safer".

> > Not sure if limiter is an answer for that at all.
> 
> Of course not. Many peaks should be removed before they reach the limiter that is the last thing in the chain.

Imagine such cases for compositions:
  1. Volume peaks at about -3 dB with 12 dB drops.
  2. Volume peaks at -12 dB with -48 dB drops.
  3. Volume generally peaks at -6 dB, drops to -36 dB but has occassional short spikes to 0 dB.

I would like to boost everything +12 dB, so they all would peak at 0 dB without sounding dull or scratchy. I hoped that Calf's ASC and "min. release" magic would control release time and shorten it only when it's safe. And, so far, I thought that it did give quite pleasant sound in such case (although, for music I generally try not to smash against limiter) but now LSP is a real contender.

But would you be confident to do something like that ? Or is doing that a horrible anti-musical blasphemy ?

> > Why not use common zoom slider of universal view widget that would scale graph inside static frame from 12 dB to 144 dB
> 
> Because it requires additional UI tunings which are not always trivial. Each line on the graph is individual widget which requires control. And, again, I don't find it useful. If you use limiter to control more than 48 dBFS gain range, then, probably, there's something wrong with your mix and it should be updated before you apply the limiter.

I'm not using it for creating something, only playback, mostly to make everything more pronounced on lower volume and make my headphones sound less dull in comparison to straight playback. Compressor and limiter mostly come into play on videos: live talking which is not normalized and/or under-compressed and full 144 dB range of movies (which is how they are supposed to be released in all BluRays and "osH-Dee" streaming) because I, obviously, not going to be blasting nowhere near that. On those, "loud" parts of volume often is around -36 dB with rare peaks to 0 dB for some annoying screaming, exploding and such. You're not going to tell every youtuber and every studio that "60 dB is enough for everyone", right ?

I'm sure it's not an easy feat to integrate such scale to every customized graph. But it's a real world application, 144 dB is not just theory and making musical solos is not the only audio application. Even then, from what I gathered, mixing and mastering engineers are almost always at odds with performers because of such contradicting biases. And in that debate engineers have won with "144 dB encoding range with normalization & compression for 0 dB and 83 dB SPL playback target should be used". But a lot of things are not heavily normalized and/or compressed, if at all, and no one is going to target lower, real playback SPL, that's the business of playback system.

> No worries, limiter plugin used in PulseEffects is the one by Calf.

Huh, I though that it was generic LV2 host, capable of running any plugin. Weird. There is no way to launch LV2 plugins directly on PA ? Personally, I think I will soon fully migrate on [PW](https://gitlab.freedesktop.org/pipewire/pipewire/) with PA & JACK emulation but still…

**Comment by Digitalone1** [comment]
> Huh, I though that it was generic LV2 host, capable of running any plugin. Weird. There is no way to launch LV2 plugins directly on PA ? Personally, I think I will soon fully migrate on [PW](https://gitlab.freedesktop.org/pipewire/pipewire/) with PA & JACK emulation but still…

No, it's not capable of running any plugin, but has the main ones. AFAIK, PA can directly launch only LADSPA plugins, not LV2.

**Comment by sadko4u** [comment]
> But would you be confident to do something like that ? Or is doing that a horrible anti-musical blasphemy ?

I would like to add some downward compressor with soft knee and small ratio before the final signal goes to the limiter. The less work to the limiter is, the more natural sound we get. But it's custom preferences of each individual person.

> You're not going to tell every youtuber and every studio that "60 dB is enough for everyone", right

I won't tell anything to youtuber that doesn't care about the quality of the produced sound. I'll just go away.

> There is no way to launch LV2 plugins directly on PA

No, there's a special **gstreamer** plugin format: https://gstreamer.freedesktop.org/documentation/plugin-development/index.html?gi-language=c

Maybe once I'll add support of them. But currently I have other development priorities.

**Comment by wwmm** [comment]
Just to clarify a few things PulseEffects approach is different from the one built-in in Pulseaudio to load LADSPA plugins. I think Pulseaudio has its own minimal LADSPA plugin wrapper. What I did in PulseEffects was using a GStreamer pipeline to handle plugins. GStreamer wrapper is capable of loading LADSPA plugins and most of the LV2 plugins. A few of them do not load because of missing LV2 extensions in GStreamer's wrapper. It also can not load the plugin built-in graphical interface.

GStreamer pipeline is dynamic and I can inserting and removing plugins on the fly as well as changing their position in the pipeline. But as there is no support for plugins interface I had to follow a path where each plugin has a custom interface written specifically for it. Nowadays there are also translations on top of them. So becoming a generic host like Ardour without losing consistency will be hard.

Another difficulty comes from using GSettings to handle presets. Most of the time it is very easy to use and convenient because gobject binding can be used on them. But the configuration keys and their default values are defined in a xml file that has to be compiled before the application is executed. So supporting arbitrary plugins at runtime may not even be possible.

Honestly I do not mind these shortcomings because becoming some kind of "Ardour for Pulseaudio" was never the intention. The idea was to have an advanced system wide audio tweaking tool without having to bother with Jack server configuration and plugins input/output connection.  But I understand that from a few users point of view this will be disappointing.
